src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,208 @@
+        /// </summary>
+        private void EnqueueComputeDataModelTask()
+        {
+            // 'true' value is unused. This just signals to the queue that we have work to do.
+            _computeDataModelQueue.AddWork(true);
+        }
+
+        /// <summary>
+        /// Makes the LSP document symbol request and creates the data model.
+        /// </summary>
+        private async ValueTask<DocumentSymbolDataModel?> ComputeDataModelAsync(ImmutableSegmentedList<bool> _, CancellationToken cancellationToken)
+        {
+            // Jump to the UI thread to get the currently active text view. This cancellation token controls the entire DocumentOutlineControl
+            // so if we are closed/cancelled on the UI thread, when this jumps back to the UI thread, it will auto-cancel and won't continue
+            // further. We only get to the code below if the control is still in an active state.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return null;
+
+            var currentSnapshot = activeTextView.TextSnapshot;
+            var textBuffer = activeTextView.TextBuffer;
+
+            var filePath = GetFilePath();
+            if (filePath is null)
+                return null;
+
+            // Ensure we switch to the threadpool before calling ComputeModelAsync. It ensures that fetching and processing the document
+            // symbol data model is not done on the UI thread.
+            await TaskScheduler.Default;
+
+            var model = await ComputeModelAsync().ConfigureAwait(false);
+
+            // The model can be null if the LSP document symbol request returns a null response.
+            if (model is not null)
+                EnqueueFilterAndSortDataModelTask();
+
+            return model;
+
+            async Task<DocumentSymbolDataModel?> ComputeModelAsync()
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                    textBuffer, _languageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+                // If there is no matching LSP server registered the client will return null here - e.g. wrong content type on the buffer, the
+                // server totally failed to start, server doesn't support the right capabilities. For C# we might know it's a bug if we get a null
+                // response here, but we don't know that in general for all languages.
+                if (response is null)
+                    return null;
+
+                var responseBody = response.ToObject<DocumentSymbol[]>();
+                if (responseBody is null)
+                    return null;
+
+                return DocumentOutlineHelper.GetDocumentSymbolDataModel(responseBody, currentSnapshot);
```

Comment:
Why do we know that currentSnapshot is the right snapshot here? We yielded the UI thread between when we acquired it and here, so it could have changed?

Reply:
Oh it's absolutely correct that we want to capture the snapshot, but the problem is the buffer could have changed between when you ask for the snapshot originally, and when you submit the LSP request. Does the LSP API give you a way for it to return which snapshot it operated on?